universal-application-tool-0.0.1/app/services/export/ExporterService.java

```diff
@@ -150,12 +150,20 @@ public String exportCsv(CsvExporter csvExporter, ImmutableList<Application> appl
     try {
       OutputStream inMemoryBytes = new ByteArrayOutputStream();
       Writer writer = new OutputStreamWriter(inMemoryBytes, StandardCharsets.UTF_8);
+      HashMap<Long, ProgramDefinition> programDefinitions = new HashMap<>();
       for (Application application : applications) {
+        Long programId = application.getProgram().id;
+        if (!programDefinitions.containsKey(programId)) {
+          try {
+            programDefinitions.put(programId, programService.getProgramDefinition(programId));
+          } catch (ProgramNotFoundException e) {
+            throw new RuntimeException("Cannot find a program that has applications for it.", e);
+          }
+        }
+        ProgramDefinition programDefinition = programDefinitions.get(programId);
+
```

Comment:
suggestion (non-blocking, if-minor): this could would be a bit quicker to read if this caching logic could be encapsulated in a private method. The `programDefinitions` cache variable could initialize in the c'tor and L155-163 become:
```java
ProgramDefinition programDefinition = getProgramDefinition(programId);
```

Where `getProgramDefinition` is a private method that implements the caching logic.

Reply:
SG.  I'm kinda in the learn/document phase of things, then I was going to take a clean up pass.